### ISSUE TITLE
Add option to check for bit rot during backup creation

### ISSUE THREAD

**Issue by oysols** [restic/restic#805]
There is currently no way of knowing if the original files deteriorate.

`restic backup` skips files based on mod time. Bit rot will not be detected at all, but the backup will contain the original file. If the owner modifies the file without discovering that the file is degraded, or `restic backup --force` is run, restic will happily backup the new file. At the time when the owner discovers that the file is degraded, it will be very hard to find and recover the correct bits. Worst case scenario, the file is lost due to backup pruning.

`restic backup --force` hashes all files, regardless of mod time. Bit rot will result in a new hash, and restic will interpret this as a normal modification and backup the file. The owner will be none the wiser, and will similarly have problems when the bit rot is detected.

I suggest implementing a third option, similar to `restic backup --force` that checks mod time to determine if the file has been changed intentionally. If the file is modified without a modified mod time it would log or err out.

(This could also be extended to verify other meta data, such as checking that the modified date is not earlier than the one recorded in the previous backup.)

This might be out of the scope of a backup program, but if we are already hashing the files, and recording the previous hashes, most of the work is already done.

**Comment by zcalusic** [comment]
You're completely right about that @oysols. And I don't think it is out of scope of restic, to check full integrity of the data it saved.

I only think this should be implemented via check command, check --full or similar. But, a good idea, definitely. 👍

**Comment by fd0** [comment]
Great idea, implementation will need to be discussed. Thanks for reporting this!

**Comment by applestump** [comment]
Great idea! +1

**Comment by ehzastrow** [comment]
+1

**Comment by bjoe2k4** [comment]
+1

**Comment by awehrfritz** [comment]
+1

**Comment by midnightmagic** [comment]
Intermediate functionality which would support this could be a dump of file hashes which users can just feed to sha256sum -c; or even just a textual retrieval of hash chunks so users like myself can write our own utilities to verify against the chunks. Use case is verifying file system files after a crash or a detected drive failure and data restoration on backups which are larger than half the size of the underlying storage and/or over a network where network hashing would be unnecessarily slow.

Really really love restic. With even just a pathway to a sum-based way to verify client drive contents it would essentially be feature-complete for someone like myself.

Much obliged for all the work you all have done. Wonderful software.

**Comment by midnightmagic** [comment]
There's a secondary utility of a more user-accessible interface to the hash tree structure on a file-and-directory basis might also be for generating existence proofs for timestamping in something like OpenTimestamp. A restic backup having extra utility for proving the historical existence of the backed-up contents would be pretty spectacular. :-)

**Comment by scvnc** [comment]
Found this thread while searching for ideas on how to do bitrot detection without needing to be an enterprise sysadmin with an expensive NAS.

I concur, if hashing is already being done then most of the work is already done :)

**Comment by dennisverspuij** [comment]
Bitrot detection by the `diff` command requires both sides to be a backup, and the right side a `--force`'d backup.
It would be great if `diff` would support the right side to be the local filesystem, diffing directly without wasting backup storage and time for a full backup just to check for bitrot. Especially when backup storage charges for minimum retention of used blocks.

**Comment by christaikobo** [comment]
@dennisverspuij maybe I am stupid...Can we simple compare 2 versions of the same backup task? Like the snapshot from yesterday and today.

**Comment by dennisverspuij** [comment]
> [@dennisverspuij](https://github.com/dennisverspuij) maybe I am stupid...Can we simply compare 2 versions of the same backup task? Like the snapshot from yesterday and today.

No because Restic only decides to backup new file contents based on whether it detects changes, and it does this using features and metadata of the underlying filesystem (e.g. block parity, checksums, timestamps, or whatsoever), not by comparing actual contents against previous backup (that would waste a lot of time and bandwith). But the latter is actually needed to detect bit rot that the filesystem and metadata is not aware of.

**Comment by christaikobo** [comment]
> > [@dennisverspuij](https://github.com/dennisverspuij) maybe I am stupid...Can we simply compare 2 versions of the same backup task? Like the snapshot from yesterday and today.
> 
> No because Restic only decides to backup new file contents based on whether it detects changes, and it does this using features and metadata of the underlying filesystem (e.g. block parity, checksums, timestamps, or whatsoever), not by comparing actual contents against previous backup (that would waste a lot of time and bandwith). But the latter is actually needed to detect bit rot that the filesystem and metadata is not aware of.

I thought `--force` would ignore change detection and rescan all files, (though how restic does it is unknown to me) which solves the problem you were mentioning?

**Comment by dennisverspuij** [comment]
Oh I see what you mean, I could create a new backup with --force and then compare the last two snapshots, but that would double the occupied backup storage space. I would like to be able to compare last snapshot against the filesystem instead by actual byte by byte comparison.

**Comment by christaikobo** [comment]
> Oh I see what you mean, I could create a new backup with --force and then compare the last two snapshots, but that would double the occupied backup storage space. I would like to be able to compare last snapshot against the filesystem instead by actual byte by byte comparison.

I think with `--force` restic rescans but not reupload all files. Dedupe is in theory still working. 

But again the details are fuzzy.

**Comment by dennisverspuij** [comment]
> > Oh I see what you mean, I could create a new backup with --force and then compare the last two snapshots, but that would double the occupied backup storage space. I would like to be able to compare last snapshot against the filesystem instead by actual byte by byte comparison.
> 
> I think with `--force` restic rescans but not reupload all files. Dedupe is in theory still working.
> 
> But again the details are fuzzy. The document has not mentioned exactly how rescan works.
> 
> Based on some other issues they don't checksum original files, checksum is only applied for pack files.
> 
> I think sometimes user needs to have some faith, unless they can and are willing to read all the source code...(maybe not ideal but it's life)

I forgot what this issue is about, when a file contains bitrot `restic backup --force` will just back it up as a new version with same metadata as previous. Having a way to have restic report this, or even better abort creating the snapshot with corrupt content at all, is the only way to catch bitrot, before pruning older shapshots that contain valid contents. But even better like second comment would be a `check --full` command that would diff the state of snapshotted files against source as equal, changed, missing or corrupt.

**Comment by christaikobo** [comment]
> > > Oh I see what you mean, I could create a new backup with --force and then compare the last two snapshots, but that would double the occupied backup storage space. I would like to be able to compare last snapshot against the filesystem instead by actual byte by byte comparison.
> > 
> > 
> > I think with `--force` restic rescans but not reupload all files. Dedupe is in theory still working.
> > But again the details are fuzzy. The document has not mentioned exactly how rescan works.
> > Based on some other issues they don't checksum original files, checksum is only applied for pack files.
> > I think sometimes user needs to have some faith, unless they can and are willing to read all the source code...(maybe not ideal but it's life)
> 
> I forgot what this issue is about, when a file contains bitrot `restic backup --force` will just back it up as a new version with same metadata as previous. Having a way to have restic report this, or even better abort creating the snapshot with corrupt content at all, is the only way to catch bitrot, before pruning older shapshots that contain valid contents. But even better like second comment would be a `check --full` command that would diff the state of snapshotted files against source as equal, changed, missing or corrupt.

I read through some more documents and now I have a better understanding of the details. The docs cover it well I just didn't find them.

Restic will cut files into chunks and perform checksum on these chunks and then pack them into pack files, the checksum of chunks are stored to be compared in the future. It uses uses Rabin Fingerprints for implementing this Content Defined Chunking, which is a deterministic algorithm so if the file is the same the chunks are always the same. 

The current method does work, if you upload and then compare these chunk files and find a chunk from a file with identical metadata as before but different checksum, it will be reported as bitrot, and you can recover the original file from an earlier snapshot. 

The `check --full` you proposed could work as well I think, maybe a dry run that only calculates chunk checksum then comparing them to a existing snapshot.
